### PR TITLE
C++ SDK refactor and add unit tests

### PR DESCRIPTION
**Pull request by halvfigur**
* Use ptr to context rather than reference
* Use call_token_t as soon as possible
* Rename call_token_t -> call_id_t
* Decouple ReplyGatherer and CallContext
* Minimize dependency on <chrono> use int64_t timestamps

### CODE REVIEW

File: src/sdk/cpp/lib/src/iotea.cpp [hunk 1]
```diff
@@ -200,27 +195,24 @@ Event Event::FromJson(const json& j) {
...
 OutgoingCall::OutgoingCall(const std::string& talent_id, const std::string& channel_id, const std::string& call_id,
                            const std::string& func, const json& args, const std::string& subject,
-                           const std::string& type)
+                           const std::string& type, int64_t when)
     : talent_id_{talent_id}
     , channel_id_{channel_id}
     , call_id_{call_id}
     , func_{func}
-    , args_{args}
+    , args_(args)
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/src/iotea.cpp:208]: What is the reason to use this type of bracket "(" here?

File: src/sdk/cpp/lib/src/iotea.cpp [hunk 2]
```diff
@@ -293,22 +285,25 @@ CallToken EventContext::Call(const Callee& callee, const json& args, const int64
...
+    return std::chrono::duration_cast<std::chrono::milliseconds>(now).count();
+}
+
 //
 // CallContext
 //
 CallContext::CallContext(const std::string& talent_id, const std::string& channel_id, const std::string& feature,
-                         const Event& event, call_handler_ptr call_handler, publisher_ptr publisher, uuid_generator_func_ptr uuid_gen)
-    : EventContext{talent_id, channel_id, event.GetSubject(), event.GetReturnTopic(), call_handler, publisher, uuid_gen}
+                         const Event& event, reply_handler_ptr reply_nandler, publisher_ptr publisher, uuid_generator_func_ptr uuid_gen)
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/src/iotea.cpp]: nandler?

File: src/sdk/cpp/lib/src/iotea.cpp [hunk 3]
```diff
@@ -487,8 +512,8 @@ Talent::Talent(const std::string& talent_id)
     : schema_{schema::Talent{talent_id}}
     , talent_id_{talent_id} {}
 
-void Talent::Initialize(call_handler_ptr call_handler, context_generator_func_ptr context_gen, uuid_generator_func_ptr uuid_gen) {
-    call_handler_ = call_handler;
+void Talent::Initialize(reply_handler_ptr reply_nandler, context_generator_func_ptr context_gen, uuid_generator_func_ptr uuid_gen) {
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/src/iotea.cpp]: nandler?

**Review by JochenKienzle — COMMENTED** [review]
I have only some small findings and I have one question: What was the intention to replace some references with pointers?
(Beside that the integration test failed, but this has to be checked in general)

File: src/sdk/cpp/lib/src/iotea.cpp (hunk 1 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/src/iotea.cpp:208]: If nlohman::json values are initialized with curly braces they get wrapped in an array with we don't want in this case.

File: src/sdk/cpp/lib/src/iotea.cpp (hunk 2 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/src/iotea.cpp]: You sure have an eye for detail :) Will fix.

File: src/sdk/cpp/lib/src/iotea.cpp (hunk 3 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/src/iotea.cpp]: Will fix.

**Review by JochenKienzle — APPROVED** [review]
Thanks for the fixes and clarification.
No further findings

**Pull request merged**